### PR TITLE
TRUNK-6010: HibernatePatientDAOTest should test getting duplicated patients by identifier and includeVoided

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
This PR fixes `HibernatePatientDAO.getDuplicatePatientsByAttributes` method by sorting passed attributes before building an SQL query. It also adds more unit tests for `identifier` and `includeVoided` attributes. It also adds random order to passed attributes in tests to make sure they aren't flaky.

## Issue I worked on
see https://issues.openmrs.org/browse/TRUNK-6010

## Checklist: I completed these to help reviewers :)
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 1]
```diff
@@ -435,19 +435,42 @@ public PatientIdentifierType getPatientIdentifierType(Integer patientIdentifierT
 	private String getDuplicatePatientsSQLString(List<String> attributes) {
 		StringBuilder outerSelect = new StringBuilder("select distinct t1.patient_id from patient t1 ");
 		final String t5 = " = t5.";
-		
+
+		List<String> sortedAttributes = new ArrayList<>();
+		// sort attributes: patient field names go first, person field names go second etc.
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: Isn't this a brittle implementation? Where sql query passes or fails depending on order of parameters?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: I agree it's not the best implementation but I didn't want to mess with the legacy code for now. The solution above fixes the method. However, we could create another ticket to fully refactor it.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: Why not use this ticket?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: I'm just not sure how much time it's gonna take, but I could try to do it here :)

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: Give it a shot and see how it goes. If it turns out to take long, we can then leave it.

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:440]: Got it

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa I turned out to be simpler than I thought :)
In this condition: `if (!outerSelect.toString().contains("person"))` it was wrongly being matched when `outerSelect` contained `person` in any way, for example in `person_name` which was obviously wrong. We only want to look for whole `person` words. That's why now I'm splitting `outerSelect` by whitespace and then checking if any element is equal to `person`. I've also added more tests that are using random order of attributes to be sure that this code work's perfectly in 100% cases.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java [hunk 2]
```diff
@@ -176,10 +179,111 @@ public void getDuplicatePatientsByAttributes_shouldNotReturnPatientsWithDuplicat
...
+		person2.addName(new PersonName("Ioan", "Theo", "Fletcher"));
+		person2.setGender("M");
+		person2.setBirthdate(Date.valueOf("2021-06-26"));
+		person2 = hibernatePersonDAO.savePerson(person2);
+		Patient patient2 = new Patient(person2);
+		patient2.addIdentifier(new PatientIdentifier("101X", null, null));
+		patient2.setVoided(true);
+		hibernatePatientDao.savePatient(patient2);
+
+		// update voided patients, doesn't update them without this line
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java:243]: This method call does not look like the correct solution for updating voided patients.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java [hunk 3]
```diff
@@ -176,10 +179,111 @@ public void getDuplicatePatientsByAttributes_shouldNotReturnPatientsWithDuplicat
...
+		hibernatePatientDao.getAllPatients(true);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java:279]: This method call does not look like the correct solution for updating voided patients.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java (hunk 2 above)
Comment (BartlomiejRasztabiga) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java]: Yeah, I just needed to flush the DB session to persist voiding patients, which this line actually did. I've replaced that with `Context.flushSession()`

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java (hunk 3 above)
Comment (BartlomiejRasztabiga) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java]: See https://github.com/openmrs/openmrs-core/pull/3814#discussion_r664114612

**Pull request merged**